pkg/worker/worker.go

```diff
@@ -0,0 +1,179 @@
+	// Stop a running Job which kills a running process.
+	//    - ID: Job identifier
+	// It returns the execution error encountered.
+	Stop(jobID string) (err error)
+	// Query a Job to check the current status.
+	//    - ID: Job identifier
+	// It returns process status and the execution error
+	// encountered.
+	Query(jobID string) (status *Status, err error)
+	// Streams the process output.
+	//    - ctx: context to cancel the log stream
+	//    - ID: Job identifier
+	// It returns read chan to stream process stdout/stderr and the
+	// execution error encountered.
+	Stream(ctx context.Context, jobID string) (logchan chan string, err error)
+}
+
+// NewWorker creates a new Worker instance.
+func NewWorker(config conf.Config) Worker {
+	return &worker{
+		logger: log.NewLogger(config),
+		jobs:   make(map[string]*Job),
+	}
+}
+
+// worker implementation.
+type worker struct {
+	// logger is responsible to handle the
+	// stdout and stderr of a running process
+	logger log.Logger
+	// jobs is concurrency safe map to store
+	// the requested jobs
+	jobs map[string]*Job
+	// mtx to control jobs concurret access
+	mtx sync.RWMutex
+}
+
+// Start runs a Linux single command with arguments.
+// If the command runs with success, a Job identifier will be returned.
+// A log file will be created with the Job id name to capture the stdout and stderr
+// of a running process.
+// To get the process status, the Job request will be stored in memory,
+// and a goroutine will be launched to update the job status when the process is finished.
+func (w *worker) Start(command Command) (string, error) {
+	cmd := exec.Command(command.Name, command.Args...)
+	jobID := uuid.NewString()
+	logfile, err := w.logger.Create(jobID)
+	if err != nil {
+		return jobID, err
+	}
+	// redirect the stdout and stderr to the log file
+	cmd.Stdout = logfile
+	cmd.Stderr = logfile
+	if err = cmd.Start(); err != nil {
+		w.logger.Remove(jobID)
+		return jobID, err
+	}
+	// create and store the job
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+	job := Job{ID: jobID, Cmd: cmd, Status: &Status{Pid: cmd.Process.Pid}}
+	w.jobs[jobID] = &job
+	// update the job status in background
+	go func() {
+		if err := job.Cmd.Wait(); err != nil {
+			logger.Printf("Command execution fails, %v", err)
+		}
+		// update the job status with information about
+		// the exited process
+		status := Status{
+			Pid:      job.Cmd.ProcessState.Pid(),
+			ExitCode: job.Cmd.ProcessState.ExitCode(),
+			Exited:   job.Cmd.ProcessState.Exited(),
+		}
+		w.mtx.Lock()
+		job.Status = &status
+		w.mtx.Unlock()
+	}()
+	return jobID, nil
+}
+
+// Stop terminates a running Job gracefully sending a SIGTERM to the process.
+// If the job doesn't exitis an error will be returned.
+func (w *worker) Stop(jobID string) error {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	job, err := w.getJob(jobID)
+	if err != nil {
+		return err
+	}
+	if err := job.Cmd.Process.Signal(syscall.Signal(0)); err != nil {
```

Comment:
What does this do?

Previously, I was referring to the fact that you could simply check the job status here to avoid sending a signal to the finished job.

Reply:
kill -0 pid is a common way to look for pid existence on POSIX. But, I got your point, I don't need to reach the OS.